src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -614,6 +622,9 @@ public bool Remove(TKey key)
                         {
                             entries[last].next = entries[i].next;
                         }
+
+                        value = entries[i].value;
```

Comment:
@mikedn hmm. A few questions, mainly as a learning for me...

> Duplicating the code might be the only option that doesn't impact performance. But it too has its cost since we're looking at a generic type.
1. Are we referring to maintenance cost? But then what is special in context of generic type... Seems there is more here... Please share.

> Using a branch to avoid the unnecessary copy might be a reasonable compromise.
2. In non-duplication way, I'm not able to understand how we could avoid assignment at line 626, since we have to squirrel off the value before being overwritten. I think I am not able to understand what is meant by _branch_ here. Appreciate your inputs.

3. I was curious why the C# compiler would not inline entire `Remove(key, out value)` into the `Remove(key)` and distill off the unnecessary copies via intermediate code optimizations. For long I have always wondered why the chained constructors also won't inline (as seen in IL for even Release mode).

Performance data points I will collect
a. Old Remove implementation
b. Nested Remove implementation as in the PR with code re-org.
c. Branched version (but I need to be sure I understand what is meant by _branch_ in this context).

Thanks!